cdap-spark-core/src/main/scala/co/cask/cdap/app/runtime/spark/DefaultSparkExecutionContext.scala

```diff
@@ -220,20 +238,40 @@ class DefaultSparkExecutionContext(runtimeContext: SparkRuntimeContext,
     }, false)
   }
 
-  private def configureStreamInput(configuration: Configuration, streamId: Id.Stream,
-                                   startTime: Long, endTime: Long): Configuration = {
-    val streamConfig = runtimeContext.getStreamAdmin.getConfig(streamId)
+  private def configureStreamInput(configuration: Configuration, streamId: StreamId, startTime: Long,
+                                   endTime: Long, formatSpec: Option[FormatSpecification]): Configuration = {
+    val streamConfig = runtimeContext.getStreamAdmin.getConfig(streamId.toId)
     val streamPath = StreamUtils.createGenerationLocation(streamConfig.getLocation,
                                                           StreamUtils.getGeneration(streamConfig))
 
     StreamInputFormat.setTTL(configuration, streamConfig.getTTL)
     StreamInputFormat.setStreamPath(configuration, streamPath.toURI)
     StreamInputFormat.setTimeRange(configuration, startTime, endTime)
-    StreamInputFormat.inferDecoderClass(configuration, classOf[StreamEvent])
-
+    // Either use the identity decoder or use the format spec to decode
+    formatSpec.fold(
+      StreamInputFormat.inferDecoderClass(configuration, classOf[StreamEvent])
+    )(
+      spec => StreamInputFormat.setBodyFormatSpecification(configuration, spec)
+    )
     configuration
   }
 
+  private def recordStreamUsage(streamId: StreamId): Unit = {
+    val oldStreamId = streamId.toId
+
+    // Register for stream usage for the Spark program
+    val oldProgramId = runtimeContext.getProgram.getId
+    val owners = List(oldProgramId)
+    try {
+      runtimeContext.getStreamAdmin.register(owners, oldStreamId)
+      runtimeContext.getStreamAdmin.addAccess(new Id.Run(oldProgramId, getRunId.getId), oldStreamId, AccessType.READ)
+    }
+    catch {
+      case e: Exception =>
+        LOG.warn("Failed to registry usage of {} -> {}", streamId, owners, e)
```

Comment:
also nit: do you think the order should be ``owners -> streamId`` in this message?